candidate.go

```diff
@@ -188,13 +189,19 @@ func (c *Candidate) recvLoop() {
 			}
 
 			continue
-		} else {
-			err = c.agent.run(func(agent *Agent) {
-				agent.noSTUNSeen(c, srcAddr)
-			})
-			if err != nil {
-				log.Warnf("Failed to handle message: %v", err)
+		}
+
+		var isValidRemoteCandidate uint32
+		err = c.agent.run(func(agent *Agent) {
```

Comment:
AFAIK the `run` function will only block until the given `func(*Agent)` function is passed to the `taskLoop`. It won't actually block on the function being fully executed.

You have to block with a channel here; atomic is not good enough. That or try to follow the path of my PR and have a `runInline` function that does block.

Reply:
Fixed.

I am going to move everything that way though, I just want to change things one at a time.